spec/requests/odk/submission_spec.rb

```diff
@@ -145,4 +145,54 @@
       expect(nemo_response.children.size).to eq(1)
     end
   end
+
+  context "duplicate submission" do
+    let(:xml_values) { %w[A B C D] }
+    let!(:form1) { create(:form, :live, mission: mission, question_types: question_types) }
+    let!(:formver) { create(:form_version, code: "abc", number: "202211", form: form1) }
+    let(:r1_path) { "tmp/odk/responses/simple_response/simple_response.xml" }
+    let(:r1) do
+      create(
+        :response,
+        :with_odk_attachment,
+        xml_path: r1_path,
+        form: form1,
+        answer_values: xml_values
+      )
+    end
+    let!(:question_types) { %w[text text text text] }
+
+    it "should return created" do
+      prepare_odk_response_fixture("simple_response", form1, values: xml_values, formver: "202211")
+      r1
+      r1_original_path = Rails.root.join("tmp/odk/responses/simple_response/simple_response.xml")
+      r2_path = Rails.root.join("tmp/odk/responses/simple_response/simple_response2.xml")
+      FileUtils.cp(r1_original_path, r2_path)
+
+      upload = Rack::Test::UploadedFile.new(r2_path, "text/xml")
+      request_params = {xml_submission_file: upload, format: "xml"}
+      post(submission_path, params: request_params, headers: auth_header)
+      expect(response).to have_http_status(:created)
+    end
+
+    context "two threads trying to submit at the same time" do
+      describe "thread test", database_cleaner: :truncate do
+        it "should not create duplicates", thread: true do
+          prepare_odk_response_fixture("simple_response", form1, values: xml_values, formver: "202211")
+          r1_path = Rails.root.join("tmp/odk/responses/simple_response/simple_response.xml")
+          upload = Rack::Test::UploadedFile.new(r1_path, "text/xml")
+          request_params = {xml_submission_file: upload, format: "xml"}
+
+          expect do
+            threads = []
+            5.times do
+              threads << Thread.new { post(submission_path, params: request_params, headers: auth_header) }
+              sleep(0.5)
```

Comment:
ok, sooooo i was thinking in real life conditions that the server would not receive the request as fast as the test framework could make the threads. but maybe i'm wrong. this does fail if the sleep is not in there because we have a race condition where the response has not been saved yet, so we can't look at the checksum yet. i leave it up to you as to whether we should continue to problem solve around it. we could consider locking around user id or something?

Reply:
ok now it is not failing....

Reply:
i thought there was a race condition but there doesn't seem to be now! which makes me a lil hmm. i was also thinking we could have a unique constraint on the checksum on the database level so even with race conditions it wouldn't happen. but then we would need to address the case that another user has an exact same submission as you mentioned.

Reply:
hmm yes unless we can explicitly confirm that rails never parallelizes these requests (I think it does parallelize them), I think we need to address checkbox #2 above to handle this.

imagined scenario:
* user submits a big response that takes 120 seconds to process
* 60 seconds later, user gets bored, clicks back in the app, and tries submitting again
* rails accepts the second request, starts a parallel thread, verifies that no checksum matches yet, and works on saving it
* we end up with a duplicate

imagined mitigation strategy:
* at the very end when saving the response, double check that it's STILL not a duplicate; if it is determined to NOW be a duplicate, we need to destroy it (this logic will be safe and won't run in parallel, Rails guarantees that)
* this spec SHOULD fail until this new check exists

this new check could indeed come in the form of a uniqueness constraint such as unique [checksum, submitter ID] -- but that would require cleaning up legacy data which we aren't doing (many things would be easier if we cleaned up legacy data but I understand why it's important not to for research/consistency reasons).

so instead I think this just needs to be regular code to basically run your `duplicate?` check again at the end and if detected, destroy the duplicate instead of saving it.